Objects/cown.c

```diff
@@ -0,0 +1,323 @@
+#include "Python.h"
+#include <ctype.h>
+#include <semaphore.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include "methodobject.h"
+#include "modsupport.h"
+#include "object.h"
+#include "pycore_ast.h"
+#include "pycore_dict.h"
+#include "pycore_interp.h"
+#include "pycore_object.h"
+#include "pycore_regions.h"
+#include "pycore_pyerrors.h"
+#include "pyerrors.h"
+#include "pystate.h"
+
+typedef enum CownState {
+    Cown_RELEASED,
+    Cown_ACQUIRED,
+    Cown_PENDING_RELEASE
+} CownState;
+
+typedef struct PyCownObject {
+    PyObject_HEAD
+    atomic_int state;
+    size_t owning_thread;
+    sem_t semaphore;
+    PyObject* value;
+} PyCownObject;
+
+static PyObject *PyCown_set_unchecked(PyCownObject *self, PyObject *args);
+static PyObject *PyCown_set(PyCownObject *self, PyObject *args);
+static PyObject *PyCown_get(PyCownObject *self, PyObject *args);
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args);
+
+static void PyCown_dealloc(PyCownObject *self) {
+#ifndef NDEBUG
+    fprintf(stderr, "Cown deallocated (%p)\n", self);
+#endif
+    sem_destroy(&self->semaphore);
+
+    PyTypeObject *tp = Py_TYPE(self);
+    PyObject_GC_UnTrack((PyObject *)self);
+    Py_XDECREF(self->value);
+    self->value = NULL;
+    PyObject_GC_Del(self);
+    Py_DECREF(tp);
+}
+
+static int PyCown_init(PyCownObject *self, PyObject *args, PyObject *kwds) {
+    // TODO: should not be needed in the future
+    _Py_MakeImmutable(_PyObject_CAST(Py_TYPE(self)));
+    _Py_notify_regions_in_use();
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown created (%p)\n", self);
+#endif
+
+    sem_init(&self->semaphore, 0, 0);
```

Comment:
`sem_init` can fail. Since this is just a temporary/initial implementation of Cowns, it might be sufficient to have a simple debug print:

```suggestion
    if (sem_init(&self->semaphore, 0, 0)) {
        eprintf("ERROR: Failed to create a semaphore");
        return -1;
    }
```